coucharchive

```diff
@@ -261,8 +287,25 @@ def make_throttler_faster():
         throttler.delay /= 2 ** (1.0 / NUMBER_OF_WORKERS)
 
 
+def queue_control_function(urls):
+    while True:
+        time.sleep(0.5)  # arbitrary
+        should_slow = False
+        for url in urls:
+            couch_queue = _get_couch_system_stats(url)['run_queue']
```

Comment:
I think that it will.

Reply:
If I understand correctly, we will do a request to each server every 0.5 second. I seems a lot, no? Can this participate to the server overload?

Is it an option to call this route only every 5 seconds, but use something like `make_throttler_slower(10×)`?

Reply:
> If I understand correctly, we will do a request to each server every 0.5 second.

Indeed.

> I seems a lot, no? Can this participate to the server overload?

I was wondering too, but replications them-self seems way worse.

Reason I'm saying that is because CouchDB seems to be really quick to answer to that API event when there is huge load (otherwise it would be impractical for this use case).
I think the data are pre-calculated and just dumped to the API when called.
So the added charged are only on the TCP stack that have a bit more work to do, but all in all I think it's reasonable.

And benefits are:
- no code modification of the throttler
- quicker and smoother control loop

*10 each 5 second seems low. It means we will need 3 consecutive calls (15 seconds) to reach a 60 seconds throttler:
```
0s --------> 1s -----------> 20s ----------> 400s (top at 60s)
     +1           x2x10
```
Today we reach a 60 seconds throttler in 7 calls (3.5 seconds):
```
0s ---> 1s ---> 2s ---> 4s ---> 8s ---> 16s ---> 32s ---> 64s  (top at 60s)
    +1     x2      x2     ...
```

What do you think of :
- call the API each two seconds
- modification of the throller value direcly:
  - when raising up, If equals `0` then make it `30`, otherwise multiple by 20
  - also lower it quicker (it's a default with the current PR code it lower slowly) by dividing by 20 or something like that.
  - maybe another smarter solution, modification of the throttler value directly but using `x = run_queue * 5` for exemple

Reply:
Very interesting!

> Today we reach a 60 seconds throttler in 7 calls (3.5 seconds):

I don't get that: today, we don't call `make_throttler_slower()` every *n* seconds: it's called only when there is an error, isn't? So it doesn't take "3.5 seconds", in average it should be more like ~ 30 seconds, no?

Anyway:
- +1 to call the API every two seconds
- > If equals 0 then make it 30, otherwise multiple by 20

  What?!? :smile: It means if  `make_throttler_slower()` is called twice, it will delay all replications by `30 × 20 = 600` seconds?

- the other solution (using the run queue value) is smart indeed, it can be a good one!

Reply:
You may have been right, I do not remember sorry.
The code on which this PR is changed a bit, so I will resolve the conversation.
If the issue is still in the new code don't hesitate to open a new one!